From 642b7dc7739e99397047192b567cb3914a9c23d2 Mon Sep 17 00:00:00 2001
From: Julian Andres Klode <jak@debian.org>
Date: Mon, 20 Jun 2011 17:30:40 +0200
Subject: [PATCH] lib: common: Fix va_list double-use (Closes: #630824)

If __get_formatted_string() is asked to format a string
of more than 4096 characters, a second call to vsnprintf()
is issued, reusing the same va_list as in the first
call. As this va_list is already exhausted, a crash
occurs.

Reported-by: Julian Andres Klode <jak@debian.org>
Signed-off-by: Julian Andres Klode <jak@debian.org>
---
 cpp/lib/src/common.cpp |    7 ++++++-
 1 files changed, 6 insertions(+), 1 deletions(-)

diff --git a/cpp/lib/src/common.cpp b/cpp/lib/src/common.cpp
index 422f395..5188a0f 100644
--- a/cpp/lib/src/common.cpp
+++ b/cpp/lib/src/common.cpp
@@ -55,10 +55,15 @@ string __get_formatted_string(const char* format, va_list va)
 {
 	char formattedBuffer[4096];
 
+	va_list vb;
+	va_copy(vb, va);
+
 	auto substitutedFormat = __substitute_eee(format, errno);
 
 	auto bytesWritten = vsnprintf(formattedBuffer, sizeof(formattedBuffer),
-			substitutedFormat.c_str(), va);
+			substitutedFormat.c_str(), vb);
+
+	va_end(vb);
 
 	if ((size_t)bytesWritten < sizeof(formattedBuffer))
 	{
-- 
1.7.5.4

